File: typing/builtins.go
package typing

func builtinPopulatedTable() map[string]Type {
	table := map[string]Type{
		"print_int":     &Fun{UnitType, []Type{IntType}},
		"print_bool":    &Fun{UnitType, []Type{BoolType}},
		"print_float":   &Fun{UnitType, []Type{FloatType}},
		"print_str":     &Fun{UnitType, []Type{StringType}},
		"println_int":   &Fun{UnitType, []Type{IntType}},
		"println_bool":  &Fun{UnitType, []Type{BoolType}},
		"println_float": &Fun{UnitType, []Type{FloatType}},
		"println_str":   &Fun{UnitType, []Type{StringType}},
		"float_to_int":  &Fun{IntType, []Type{FloatType}},
		"int_to_float":  &Fun{FloatType, []Type{IntType}},
		"str_size":      &Fun{IntType, []Type{StringType}},
		"__str_equal":   &Fun{BoolType, []Type{StringType, StringType}},
		"str_concat":    &Fun{StringType, []Type{StringType, StringType}},
		"substr":        &Fun{StringType, []Type{StringType, IntType, IntType}},
		"int_to_str":    &Fun{StringType, []Type{IntType}},
		"float_to_str":  &Fun{StringType, []Type{FloatType}},
		"str_to_int":    &Fun{IntType, []Type{StringType}},
		"str_to_float":  &Fun{FloatType, []Type{StringType}},
		"get_line":      &Fun{StringType, []Type{UnitType}},
	}
	return table
}

File: codegen/testdata/builtins.ml
print_int 42;
print_bool true;
print_float 3.14;
print_str "PIYOOO";
println_bool false;
println_int 10;
println_float 3.14;
println_str "POYOOO";

println_int (float_to_int 3.14);
println_float ((int_to_float 3) +. 1.1);
println_int (str_length "aiueo");

let a = int_to_str 42 in
println_str a;
println_int (str_length a);
let a = int_to_str (-10) in
println_str a;
println_int (str_length a);

let a = float_to_str 3.14 in
println_str a;
println_int (str_length a);
let a = float_to_str (-.1.123) in
println_str a;
println_int (str_length a);

println_int (str_to_int "1234");
println_int (str_to_int "-1234");

println_float (str_to_float "3.14");
println_float (str_to_float "-1.4142");

println_int (bit_and 2 1);
println_int (bit_or 3 1);
println_int (bit_xor 3 1);
println_int (bit_lsft 1 10);
println_int (bit_rsft 256 4);
println_int (bit_inv 0);

()
